pkg/storage/endpoint/gc_states.go

```diff
@@ -0,0 +1,556 @@
+	}
+	key := path.Join(prefix, barrierID)
+	// GCBarrier is stored in ServiceSafePoint format for compatibility.
+	serviceSafePoint, err := loadJSON[*ServiceSafePoint](p.storage, key)
+	if err != nil {
+		return nil, err
+	}
+	return gcBarrierFromServiceSafePoint(serviceSafePoint), nil
+}
+
+// LoadAllGCBarriers loads all GC barriers of the given keyspace.
+func (p GCStateProvider) LoadAllGCBarriers(keyspaceID uint32) ([]*GCBarrier, error) {
+	prefix := keypath.GCBarrierPrefix()
+	if keyspaceID != constant.NullKeyspaceID {
+		prefix = keypath.KeyspaceGCBarrierPrefix(keyspaceID)
+	}
+	// TODO: Limit the count for each call.
+	_, serviceSafePoints, err := loadJSONByPrefix[*ServiceSafePoint](p.storage, prefix, 0)
+	if err != nil {
+		return nil, err
+	}
+	if len(serviceSafePoints) == 0 {
+		return nil, nil
+	}
+	barriers := make([]*GCBarrier, 0, len(serviceSafePoints))
+	for _, serviceSafePoint := range serviceSafePoints {
+		barriers = append(barriers, gcBarrierFromServiceSafePoint(serviceSafePoint))
+	}
+	return barriers, nil
+}
+
+// CompatibleLoadTiDBMinStartTS loads the minStartTS reported to etcd directly by TiDB.
+func (p GCStateProvider) CompatibleLoadTiDBMinStartTS(keyspaceID uint32) (string, uint64, error) {
+	prefix := keypath.CompatibleTiDBMinStartTSPrefix()
+	if keyspaceID != constant.NullKeyspaceID {
+		prefix = keypath.CompatibleKeyspaceTiDBMinStartTSPrefix(keyspaceID)
+	}
+	prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
+
+	// TODO: Limit the count for each call.
+	keys, values, err := p.storage.LoadRange(prefix, prefixEnd, 0)
+	if err != nil {
+		return "", 0, err
+	}
+
+	if len(keys) == 0 {
+		return "", 0, nil
+	}
+
+	var minKey string
+	var minMinStartTS uint64
+
+	for i, valueStr := range values {
+		minStartTS, err := strconv.ParseUint(valueStr, 10, 64)
+		if err != nil {
+			return "", 0, errs.ErrStrconvParseUint.Wrap(err).GenWithStackByArgs()
+		}
+		if len(minKey) == 0 || minStartTS < minMinStartTS {
+			minMinStartTS = minStartTS
+			minKey = keys[i]
+		}
+	}
+
+	// Remove prefix from the key and only keep the identifier written by TiDB.
+	if len(minKey) < len(prefix) || minKey[:len(prefix)] != prefix {
+		// This is expected to be unreachable.
+		return "", 0, errors.Errorf("unexpected internal error: loading TiDB min start ts but got mismatching key prefix, expected prefix: %s, got key: %s", prefix, minKey)
+	}
+	minKey = minKey[len(prefix):]
+	return minKey, minMinStartTS, nil
+}
+
+// RunInGCStateTransaction runs a transaction for updating GC states or read a batch of GC states.
+// The atomicity is guaranteed by a "revision" key. Any non-empty write caused by the transaction will increase the
+// revision.
+// In the transaction, reads can be performed on the GCStateProvider as usual, while writes should only be performed
+// through the GCStateWriteBatch.
+func (p GCStateProvider) RunInGCStateTransaction(f func(wb *GCStateWriteBatch) error) error {
+	revisionKey := keypath.GCStateRevisionPath()
```

Comment:
Is it possible to use the revision provided by etcd?

Reply:
Yes but hard to provide it over the abstracted `kv.Base` interface.

Reply:
And it needs a single centrialized key anyway.